### ISSUE TITLE
Implement Jetpack Module Settings

### ISSUE THREAD

**Issue by oskosk** [Automattic/wp-calypso#9171]
We're a implementing Jetpack Module Settings in Calypso’s site settings page/tabs.

More tasks related to this are being tracked in #11676

Currently these are the data-related tasks that we can do before implementing the UI. 

### API related tasks

- [x] Define an authorization mechanism for making requests to a site’s REST API from Calypso on behalf of a local user. Jetpack Signatures will be used resembling the auth flow of XML-RPC requests.
- [x] Implement the defined authorization mechanism on the Jetpack Plugin side - @roccotripaldi  Automattic/Jetpack#5418 
- [x] Implement the Jetpack REST API client in .com - @roccotripaldi  - D3527-code
- [x] Implement the endpoint in the .com side - @tyxla, @roccotripaldi - D3588-code
- [x] Implement a Jetpack REST API endpoint for reading Jumpstart - @tyxla - Automattic/Jetpack#5813
- [x] Update the `/settings` readable JSON API endpoint to retrieve all settings - @tyxla - Automattic/Jetpack#6002
- [x] Add a specific parameter for triggering Jetpack WP REST API authentication - @nylen - D3926-code

### Calypso API-client tasks

- [x] Have a wpcom.js method for reading and updating Jumpstart status
  - [x] `getJetpackJumpstart` and `updateJetpackJumpstart` which use the site’s REST API are introduced in #9868 (@tyxla)
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5813 are merged; remove the TODO comment - #10612 (@tyxla)
- [x] Have a wpcom.js method for activating a Jetpack module - [jetpackModulesActivate()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/wpcom-undocumented/lib/undocumented.js#L81) works. But uses the XML-RPC API.
  - [x] `jetpackModuleActivate()` does it via the site’s REST API - #10075 - @tyxla
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5418 are merged; remove the TODO comment. - #10612 (@tyxla)
- [x] Have a wpcom.js method for deactivating a Jetpack module - [jetpackModulesDeactivate()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/wpcom-undocumented/lib/undocumented.js#L94) works. But uses the XML-RPC API.
  - [x] `jetpackModuleDeactivate()` does it via the site’s REST API - #10076 - @tyxla
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5418 are merged; remove the TODO comment. - #10612 (@tyxla)
- [x] Have a wpcom.js method for reading a Jetpack module's activation state - [jetpackModules()](https://github.com/Automattic/wp-calypso/blob/master/client/lib/wpcom-undocumented/lib/undocumented.js#L68) works. But uses the XML-RPC API.
  - [x] `getJetpackModules()` does it via the site’s REST API - #10077 - @tyxla
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5418 are merged; remove the TODO comment. - #10612 (@tyxla)
- [x] Have a wpcom.js method for bulk reading Jetpack module settings - `fetchJetpackSettings()` does it via the site’s REST API - #10123 - @tyxla
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5418 are merged; remove the TODO comment. - #10612 (@tyxla)
- [x] Have a wpcom.js method for bulk updating Jetpack module settings - `updateJetpackSettings()` does it via the site’s REST API - #10123 - @tyxla
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5418 are merged; remove the TODO comment. - #10612 (@tyxla)
- [x] Have a wpcom.js method for retrieving Jetpack connection data - `getJetpackConnectionStatus()` does it via the site’s REST API - #10002 - @tyxla
  - [x] Do a final test after D3588-code, D3926-code and Automattic/Jetpack#5418 are merged; remove the TODO comment. - #10612 (@tyxla)

### Module activation/deactivation tasks

- [x] Create action types, action creators, and reducer for **activating a Jetpack module** - @oskosk - #9245
- [x] Create action types, action creators and reducer for **deactivating a Jetpack module** - @tyxla - #9418 
- [x] Create action types , action creators and reducer for **reading Jetpack modules activation state** - @oskosk - #9440
- [x] Implement notices on module activation and deactivation - @tyxla - #10583

### Module general tasks
- [x] Create a `QueryJetpackModules` query component  - @tyxla - #9659
- [x] Create selectors for retrieving general module data - @tyxla - #9661

### Module settings & Jumpstart tasks

- [x] Create action types, action creators and reducer for **updating a jetpack module setting**. - @tyxla - #9847 (deleted in favor of bulk settings)
- [x] Create action types, action creators and reducer for **jumpstarting a site**. - @tyxla - #9868 
- [x] Be able to read current module settings - @tyxla - #9847 (deleted in favor of bulk settings)
- [x] Create action types, action creators and reducer for **bulk reading module settings**. - @tyxla - #10123
- [x] Create action types, action creators and reducer for **bulk updating module settings**. - @tyxla - #10123
- [x] Create reducer and selectors for settings saving - @tyxla - #10694
- [x] Create a mechanism to normalize and sanitize fields - @tyxla - #10776
- [x] Implement Post by Email action, reducer and selectors - @tyxla - #10811
- [X] ~Revert #10891 once Jetpack 4.6 lands.~ Won't do as we considered its best to keep this syncing code in place anyways.

## UI tasks

**These tasks and notes are based on what's specified by p6TEKc-HQ-p2**

**Some notes on UI behavior**

* When a module is disabled (via a clear toggle like Carousel), hide module options. Not necessary, but could be nice.
* When a toggle is used, it should immediately enable that thing and show a global notice to show status of enabling that thing.
* When checkboxes are used, clicking a save settings button is required.
* A user can click a save settings button at any time even if not needed (though not when a setting is being applied). Try the buttons in Calypso to see how they work.

### General tasks

* [x] Introduce an `<InfoLink />` component for linking to module descriptions next to the module activation toggle - [Proposed here ](https://github.com/Automattic/wp-calypso/pull/9802#discussion_r91405194) - @oskosk - #10036.
* [x] Introduce a `<JetpackModuleToggle />` component to allow easy toggling of modules - @tyxla - #10429
* [x] Make `wrapSettingsForm` compatible with Jetpack Settings so we can use it for the UI cards - @tyxla - #10697

### General tab
* [x] Replace checkboxes with toggles for consistency - @tyxla - #10229

### Writing Tab

* [x] Move current Writing settings into separate cards - #9977 - @tyxla
* [x] Update the card for Custom Content Types to the new layout - #11096 - @tyxla
* [x] Implement Card for **Media** settings -  #9802.
  * Implement toggle for _Speed up your images and photon with Photon._ Has description: _Enabling Photon is required to use Tiled Galleries._
  * Implement toggle for _Transform image galleries into full-screen slideshows._
    * Implement checkbox for _Show photo metadata (Exif) in carousel, when available._
    * Implement select for _Background color_
* [x] Implement Card for **Theme enhancements** settings - #10083 - @tyxla
  * Implement toggle for _Add support for infinite scroll to your theme._
    * Implement checkbox for _Scroll infinitely (Shows 7 posts on each load)._
    * Implement checkbox for _Track each infinite Scroll post load as a page view in Google Analytics._
  * Implement toggle for _Optimize your site with a phone-friendly theme._
    * Implement checkbox for _Use excerpts instead of full posts on front page and archive pages._
    * Implement checkbox for _Hide all featured images._
    * Implement checkbox for _Show an ad for the WordPress mobile apps in the footer of the mobile theme._
* [x] Implement Card for **Publishing tools** settings. - #10663 - @tyxla
  * [x] Prepare Press This for embedding within another card - #10659 - @tyxla
  * Implement toggle for _Publish posts by email, using any device and email client._
    * Implement read-only input with copy button for _Email Address._
    * Implement button for _Regenerate Address_.
    * Move **Press This** settings into this card

### Discussion Tab

* [x] Move current Discussion settings into separate cards - @ryelle - #10167
* [x] Implement the Card for **Article** settings
  * Includes the three existing settings for **Default article settings** but as toggles.
* [x] Implement the Card for **Subscriptions** settings - @tyxla - #10815
  * Implement toggle for _Allow users to subscribe to your posts and comments and receive notifications via email._
  * Implement the checkbox for _Show a "follow blog" option in the comment form._
  * Implement the checkbox for _Show a "follow comments" option in the comment form._
   
  

### Things to have in mind (and related tasks)

- Take in account those modules that should be disabled if the site’s is in dev mode. 
  * [x] Make the development mode accessible in the Redux tree - #10002 - @tyxla
- Take in account we should **probably** not modify live settings when site is in staging.
  * [x] Make the staging mode accessible in the Redux tree - #10002 - @tyxla
- Take in account features that are only available in some of the plans.

### Notes

- Every time the word setting is used here, remember that it maps to either:
  - A single WP option.
  - A key inside a JSON object saved in a WP option

**Comment by oskosk** [comment]
Closing as every task listed here is already complete. Opened a new task list as issue #11676 as a follow-up to this
